File: snake-game/typescript/src/Direction.ts
import type { Position } from './Position.js';

export type Direction = 'Up' | 'Down' | 'Left' | 'Right';

const opposites: Record<Direction, Direction> = {
  Up: 'Down',
  Down: 'Up',
  Left: 'Right',
  Right: 'Left',
};

export function isOpposite(a: Direction, b: Direction): boolean {
  return opposites[a] === b;
}

export function move(direction: Direction, position: Position): Position {
  switch (direction) {
    case 'Up':
      return { x: position.x, y: position.y - 1 };
    case 'Down':
      return { x: position.x, y: position.y + 1 };
    case 'Left':
      return { x: position.x - 1, y: position.y };
    case 'Right':
      return { x: position.x + 1, y: position.y };
  }
}

File: snake-game/python/src/snake_game/direction.py
from __future__ import annotations

from enum import Enum

from .position import Position


class Direction(Enum):
    UP = "Up"
    DOWN = "Down"
    LEFT = "Left"
    RIGHT = "Right"

    def is_opposite(self, other: Direction) -> bool:
        return _OPPOSITES[self] == other

    def move(self, position: Position) -> Position:
        dx, dy = _DELTAS[self]
        return Position(position.x + dx, position.y + dy)


_OPPOSITES: dict[Direction, Direction] = {
    Direction.UP: Direction.DOWN,
    Direction.DOWN: Direction.UP,
    Direction.LEFT: Direction.RIGHT,
    Direction.RIGHT: Direction.LEFT,
}

_DELTAS: dict[Direction, tuple[int, int]] = {
    Direction.UP: (0, -1),
    Direction.DOWN: (0, 1),
    Direction.LEFT: (-1, 0),
    Direction.RIGHT: (1, 0),
}
